File: client/src/services/userService.js
import apiClient from "./apiClient";

export const fetchUserProfile = async (
  userId,
  includePosts,
  includeJoinedClubs
) => {
  const { data } = await apiClient.get(
    `/users/${userId}?includePosts=${includePosts}&includeJoinedClubs=${includeJoinedClubs}`
  );
  return data;
};

export const updateProfilePicture = async (userId, imageUrl) => {
  const { data } = await apiClient.put(`/users/change/${userId}`, imageUrl, {
    headers: { "Content-Type": "text/plain" },
  });
  return data;
};

export const quickSearchUsers = async (query) => {
  if (!query?.trim()) return [];
  const { data } = await apiClient.get(
    `/users/quick-search?query=${encodeURIComponent(query)}`
  );
  return data;
};

export const searchUsers = async (query) => {
  if (!query?.trim()) return [];
  const { data } = await apiClient.get(
    `/users/search?query=${encodeURIComponent(query)}`
  );
  return data;
};

File: app/src/main/java/com/tuniclubs/app/services/UserService.java
package com.tuniclubs.app.services;

import com.tuniclubs.app.dto.UserDTO;
import com.tuniclubs.app.mapper.UserMapper;
import com.tuniclubs.app.models.User;
import com.tuniclubs.app.repository.UserRepository;
import jakarta.persistence.EntityNotFoundException;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;

@Service
public class UserService {
    private final UserRepository userRepository;

    public UserService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    public UserDTO createOrUpdateUser(User user) {
        user.setUpdatedAt(LocalDateTime.now());

        if (!userRepository.existsByUserId(user.getUserId())) {
            user.setCreatedAt(LocalDateTime.now());
        } else {
            User existingUser = getUserByUserId(user.getUserId());
            if (existingUser.getImageUrl() != null && !existingUser.getImageUrl().contains("clerk.com")) {
                user.setImageUrl(existingUser.getImageUrl());
            }
            user.setPosts(existingUser.getPosts());
            user.setComments(existingUser.getComments());
            user.setLikedPosts(existingUser.getLikedPosts());
            user.setLikedComments(existingUser.getLikedComments());
        }

        User savedUser = userRepository.save(user);
        return UserMapper.toUserDTO(savedUser);
    }

    public UserDTO getUserDTO(String userId, boolean includePosts, boolean includeJoinedClubs) {
        User user = getUserByUserId(userId);
        return UserMapper.toUserDTO(user, includePosts, includeJoinedClubs);
    }

    public User getUserByUserId(String userId) {
        return userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));
    }

    public UserDTO updateProfilePicture(String userId, String imageUrl) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new EntityNotFoundException("User not found with ID: " + userId));

        String cleanImageUrl = imageUrl.replace("\"", "");
        user.setImageUrl(cleanImageUrl);
        user.setUpdatedAt(LocalDateTime.now());

        User updatedUser = userRepository.save(user);
        return UserMapper.toUserDTO(updatedUser);
    }

    public List<UserDTO> quickSearchUsers(String query) {
        if (query == null || query.trim().isEmpty()) {
            return List.of();
        }

        Pageable limit = PageRequest.of(0, 4); // Limit to 4 results
        return userRepository.quickSearchUsers(query.trim(), limit).stream()
                .map(UserMapper::toUserDTO)
                .toList();
    }

    public List<UserDTO> searchUsers(String query) {
        if (query == null || query.trim().isEmpty()) {
            return List.of();
        }

        return userRepository.searchUsers(query.trim()).stream()
                .map(UserMapper::toUserDTO)
                .toList();
    }
}
